node/src/components/small_network.rs

```diff
@@ -953,6 +884,67 @@ where
+            Event::ContractRuntimeAnnouncement(
+                ContractRuntimeAnnouncement::LinearChainBlock(_)
+                | ContractRuntimeAnnouncement::StepSuccess { .. },
+            ) => Effects::new(),
+            Event::ContractRuntimeAnnouncement(
+                ContractRuntimeAnnouncement::UpcomingEraValidators {
+                    era_that_is_ending,
+                    mut upcoming_era_validators,
+                },
+            ) => {
+                if era_that_is_ending < self.active_era {
+                    debug!("ignoring past era end announcement");
+                } else if era_that_is_ending != self.active_era {
+                    debug!("received upcoming era validators announcement, clearing limiters");
+                    self.active_era = era_that_is_ending + 1;
+                    self.active_era_validators = Default::default();
+
+                    self.incoming_limiter
+                        .update_validators(Default::default(), Default::default());
+                    self.outgoing_limiter
+                        .update_validators(Default::default(), Default::default());
+                } else {
+                    // At this point we can rely on the `active_era` being the one that is ending.
+                    self.active_era = era_that_is_ending + 1;
+
+                    // Initialize the next set of active validators with the currently active one.
+                    let mut active_validators: HashSet<PublicKey> =
+                        mem::take(&mut self.active_era_validators);
+
+                    self.active_era_validators = upcoming_era_validators
+                        .remove(&self.active_era)
+                        .unwrap_or_default()
+                        .into_keys()
+                        .collect();
+
+                    if self.active_era_validators.is_empty() {
+                        warn!("received an empty set of active era validators");
+                    }
+
+                    active_validators.extend(self.active_era_validators.iter().cloned());
+
+                    let upcoming_validators: HashSet<PublicKey> = upcoming_era_validators
+                        .remove(&(self.active_era + 1))
+                        .unwrap_or_default()
+                        .into_keys()
+                        .collect();
+
+                    debug!(
+                        %era_that_is_ending,
+                        active = active_validators.len(),
+                        upcoming = upcoming_validators.len(),
+                        "updating active and upcoming validators"
+                    );
+                    self.incoming_limiter
+                        .update_validators(active_validators.clone(), upcoming_validators.clone());
+                    self.outgoing_limiter
+                        .update_validators(active_validators, upcoming_validators);
+                }
+
+                Effects::new()
+            }
```

Comment:
This is the part that will be simplified as discussed, is it? We'll just prioritize the union of the `upcoming_era_validators` values and that's it?

Reply:
Yes, see c5720c575